kafkatest/server.go

```diff
@@ -410,6 +435,88 @@ func (s *Server) handleOffsetRequest(nodeID int32, conn net.Conn, req *proto.Off
 	return resp
 }
 
+func (s *Server) handleConsumerMetadataRequest(nodeID int32, conn net.Conn, req *proto.ConsumerMetadataReq) response {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	addrps := strings.Split(s.Addr(), ":")
+	port, _ := strconv.Atoi(addrps[1])
+
+	return &proto.ConsumerMetadataResp{
+		CorrelationID:   req.CorrelationID,
+		CoordinatorID:   0,
+		CoordinatorHost: addrps[0],
+		CoordinatorPort: int32(port),
+	}
+}
+
+func (s *Server) getTopicOffset(group, topic string, partID int32) *topicOffset {
+	pmap, ok := s.offsets[topic]
+	if !ok {
+		pmap = make(map[int32]map[string]*topicOffset)
+		s.offsets[topic] = pmap
+	}
+
+	groups, ok := pmap[partID]
+	if !ok {
+		groups = make(map[string]*topicOffset)
+		pmap[partID] = groups
+	}
+
+	toffset, ok := groups[group]
+	if !ok {
+		toffset = &topicOffset{}
+		groups[group] = toffset
+	}
+
+	return toffset
+}
+
+func (s *Server) handleOffsetFetchRequest(nodeID int32, conn net.Conn, req *proto.OffsetFetchReq) response {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	resp := &proto.OffsetFetchResp{
+		CorrelationID: req.CorrelationID,
+		Topics:        make([]proto.OffsetFetchRespTopic, len(req.Topics)),
+	}
+	for ti, topic := range req.Topics {
+		respPart := make([]proto.OffsetFetchRespPartition, len(topic.Partitions))
+		resp.Topics[ti].Name = topic.Name
+		resp.Topics[ti].Partitions = respPart
+		for pi, part := range topic.Partitions {
+			toffset := s.getTopicOffset(req.ConsumerGroup, topic.Name, part)
+			respPart[pi].ID = part
+			respPart[pi].Metadata = toffset.metadata
+			respPart[pi].Offset = toffset.offset
+		}
+	}
+	return resp
+}
+
+func (s *Server) handleOffsetCommitRequest(nodeID int32, conn net.Conn, req *proto.OffsetCommitReq) response {
```

Comment:
```
toffset := s.getTopicOffset(req.ConsumerGroup, topic.Name, part.ID)
toffset.metadata = part.Metadata
toffset.offset = part.Offset
```

That fetches the `*topicOffset` struct and since it's a pointer it updates the fields. This does save the values.